src/ruby/puppetserver-lib/puppet/server/compiler.rb

```diff
@@ -10,15 +10,43 @@ def initialize
 
       def compile(request_data)
         processed_hash = convert_java_args_to_ruby(request_data)
+        persist = processed_hash['persistence']
 
         node = create_node(processed_hash)
 
+        if persist['facts']
+          save_facts(node.facts, node.trusted_data)
+        end
+
         catalog = Puppet::Parser::Compiler.compile(node, processed_hash['job_id'])
+
+        if persist['catalog']
+          save_catalog(catalog)
+        end
+
         catalog.to_data_hash
       end
 
       private
 
+      def save_facts(facts, trusted_facts)
+          if Puppet::Node::Facts.indirection.terminus_class.to_s == "puppetdb"
+            Puppet.override({trusted_information: trusted_facts}) do
+              Puppet::Node::Facts.indirection.save(facts)
+            end
+          else
+            raise Puppet::Error, "PuppetDB not configured, not saving Facts"
+          end
+        end
```

Comment:
Soo... I think we might not want to be this restrictive. Since we are just calling into the indirector now to do the save, I think we should rely on it to do the right thing, and not raise if it can't find PDB.

@tvpartytonight ran into an issue with this when trying to use our new endpoint to use the PDB module to install PDB. The facts and catalog from that run should be saved (it's a normal agent run), but they can't be because PDB isn't present yet. This kind of issue probably isn't relevant for our current known use cases, but it will be if we ever try to use this endpoint more generally.

Please push back if you think there is value in having `persistence` mean `PDB only`.

Reply:
I'm fine with this not requiring PDB, however, I don't think the change in approach to the Indirector has an effect on that decision.

Reply:
Oh sorry, I meant as opposed to what I was originally doing when I was instantiating the PDB terminus explicitly, and we were checking settings to make sure Puppet expected that to work.

Reply:
Which I guess we are still doing for retrieval. But I guess that part makes sense?